File: zig/src/aud/soundio.zig
const RingBuffer = @import("../ring_buffer.zig").RingBuffer;
const common = @import("common.zig");
const helpers = @import("../helpers.zig");
const std = @import("std");
const c = @cImport({
    @cInclude("soundio/soundio.h");
});

const Allocator = std.mem.Allocator;
const ArrayList = std.ArrayList;
const AudError = common.AudError;
const AudioFormat = common.AudioFormat;
const ConnectContext = common.ConnectContext;
const Device = common.Device;
const Direction = common.Direction;
const PrioritizedSampleRates = common.PrioritizedSampleRates;
const expect = std.testing.expect;
const expectEqual = std.testing.expectEqual;
const expectEqualStrings = std.testing.expectEqualStrings;
const mem = std.mem;
const print = std.debug.print;
const talloc = std.testing.allocator;
const time = std.time;

// Use the SoundIo enum for formats because these enum values
// will not be transferrable across different backends.
pub const PrioritizedFormats = [_]c_int{
    c.SoundIoFormatS8,
    c.SoundIoFormatU8,
    c.SoundIoFormatS16LE,
    c.SoundIoFormatS16BE,
    c.SoundIoFormatU16LE,
    c.SoundIoFormatU16BE,
    c.SoundIoFormatS24LE,
    c.SoundIoFormatS24BE,
    c.SoundIoFormatU24LE,
    c.SoundIoFormatU24BE,
    c.SoundIoFormatS32LE,
    c.SoundIoFormatS32BE,
    c.SoundIoFormatU32LE,
    c.SoundIoFormatU32BE,
    c.SoundIoFormatFloat32LE,
    c.SoundIoFormatFloat32BE,
};

pub const DefaultLatencyMs: c_int = 0;
pub const DefaultBufferMs: c_int = 4000;

fn failed(status: c_int) bool {
    return status != 0;
}

fn index_failed(status: c_int) bool {
    return status < 0;
}

// TODO(lbayes): Work out backend_disconnect scenario outlined here:
// http://libsound.io/doc-2.0.0/backend_disconnect_recover_8c-example.html
//
// TODO(lbayes): Decide if we want to work with 'raw' devices.
//
// According to SoundIO docs found here:
//
// http://libsound.io/doc-2.0.0/structSoundIoDevice.html#afb73b9bd13e98a14a187567f508acebe
//
// And example here:
//
// http://libsound.io/doc-2.0.0/sio_record_8c-example.html
//
// Device.is_raw indicates that, "you are directly opening the
// hardware device and not going through a proxy such as dmix, PulseAudio, or
// JACK.
//
// When you open a raw device, other applications on the computer are not able
// to simultaneously access the device. Raw devices do not perform automatic
// resampling and thus tend to have fewer formats available."

const NativeGetDevice = fn ([*c]c.struct_SoundIo, c_int) callconv(.C) [*c]c.struct_SoundIoDevice;

pub const Devices = struct {
    allocator: *Allocator,
    soundio: *c.SoundIo,
    aud_devices: ArrayList(*Device) = undefined,
    sio_devices: ArrayList(*c.SoundIoDevice) = undefined,
    last_id: u8 = 0,

    pub fn init(a: *Allocator) !*Devices {
        const instance = try a.create(Devices);

        const version = c.soundio_version_string();
        print("soundio version: {s}\n", .{version});

        const soundio = c.soundio_create();
        if (soundio == null) {
            print("soundio failed to allocate\n", .{});
            return error.Fail;
        }
        print("soundio successfully created context\n", .{});

        {
            // const value = @intToEnum(c.SoundIoBackend, c.SoundIoBackendPulseAudio);
            // const status = c.soundio_connect_backend(soundio, value);
            const status = c.soundio_connect(soundio);

            if (failed(status)) {
                print("------------------------------------\n", .{});
                print("soundio failed to connect to backend\n", .{});
                // Free the soundio object before exiting
                c.soundio_destroy(soundio);
                return error.Fail;
            }
            print("soundio connected with backend: {}\n", .{soundio.*.current_backend});
        }

        c.soundio_flush_events(soundio);
        print("soundio flushed events\n", .{});

        instance.* = Devices{
            .allocator = a,
            .soundio = soundio,
            .sio_devices = ArrayList(*c.SoundIoDevice).init(a),
            .aud_devices = ArrayList(*Device).init(a),
        };

        return instance;
    }

    pub fn deinit(self: *Devices) void {
        self.sio_devices.deinit();
        // Free each device that was allocated for the devices list.
        for (self.aud_devices.items) |dev| {
            self.allocator.destroy(dev);
        }
        self.aud_devices.deinit();
        c.soundio_destroy(self.soundio);
        self.allocator.destroy(self);
    }

    fn getNativeDevices(self: *Devices, buffer: []Device, count: c_int, get_by_index: NativeGetDevice) ![]Device {
        var index: usize = 0;

        while (index < count) : (index += 1) {
            const sio_device = get_by_index(self.soundio, @intCast(c_int, index));
            const device = try self.sioDeviceToAudDevice(sio_device, index);
            buffer[index] = device.*;
        }
        return buffer[0..index];
    }

    fn sioDeviceToAudDevice(self: *Devices, sio_device: *c.SoundIoDevice, c_index: usize) !*Device {
        const id = mem.sliceTo(sio_device.id, 0);
        const native_name = mem.sliceTo(sio_device.name, 0);

        const as_int = @ptrToInt(sio_device);
        const after = @intToPtr(*c.struct_SoundIoDevice, as_int);

        print("-------------------\n", .{});
        print("soundio device id: {s}\n", .{id});
        print("soundio device ptr: {d}\n", .{as_int});
        // print("SIO DEV BEFORE {*}\n", .{sio_device});
        // print("SIO PTR INT: {} {d}\n", .{ @TypeOf(as_int), as_int });
        // print("SIO PTR AFTER: {*}\n", .{after});
        print("-------------------\n", .{});

        var device = try self.allocator.create(Device);

        device.* = Device{
            .id = id,
            .name = native_name,
            .platform_device = as_int,
        };

        // Default Device.direction is Render, update if we have a Capture device.
        if (sio_device.aim == @intToEnum(c.SoundIoDeviceAim, c.SoundIoDeviceAimInput)) {
            device.direction = Direction.Capture;
        }

        // print("ID: {s}\n", .{device.id});
        // print("NAME: {d}: {s}\n", .{ mem.len(device.name), device.name });
        try self.aud_devices.append(device);

        return device;
    }

    pub fn info(self: *Devices) []const u8 {
        return "soundio";
    }

    pub fn getDefaultDevice(self: *Devices, direction: Direction) !*Device {
        if (direction == Direction.Capture) {
            return try self.getDefaultCaptureDevice();
        } else {
            return try self.getDefaultRenderDevice();
        }
    }

    pub fn getCaptureDeviceAt(self: *Devices, index: usize) !*Device {
        const sio_device = try c.soundio_get_input_device(self.soundio, @intCast(c_int, index)) orelse error.Fail;
        return try self.sioDeviceToAudDevice(sio_device, index);
    }

    pub fn getDefaultCaptureDevice(self: *Devices) !*Device {
        const index = c.soundio_default_input_device_index(self.soundio);
        if (index_failed(index)) {
            print("soundio_default_input_device_index failed with: {}\n", .{index});
            return error.Fail;
        }
        return try self.getCaptureDeviceAt(@intCast(usize, index));
    }

    pub fn getRenderDeviceAt(self: *Devices, index: usize) !*Device {
        const sio_device = try c.soundio_get_output_device(self.soundio, @intCast(c_int, index)) orelse error.Fail;
        return try self.sioDeviceToAudDevice(sio_device, index);
    }

    pub fn getDefaultRenderDevice(self: *Devices) !*Device {
        const index = c.soundio_default_output_device_index(self.soundio);
        if (index_failed(index)) {
            print("soundio_default_output_device_index failed with: {}\n", .{index});
            return error.Fail;
        }
        return try self.getRenderDeviceAt(@intCast(usize, index));
    }

    pub fn getDevices(self: *Devices, buffer: []Device, direction: Direction) ![]Device {
        if (direction == Direction.Capture) return self.getCaptureDevices(buffer) else return self.getRenderDevices(buffer);
    }

    pub fn getCaptureDevices(self: *Devices, buffer: []Device) ![]Device {
        print("CAPTURE DEVICES\n", .{});
        const count = c.soundio_input_device_count(self.soundio);
        return self.getNativeDevices(buffer, count, c.soundio_get_input_device);
    }

    pub fn getRenderDevices(self: *Devices, buffer: []Device) ![]Device {
        print("RENDER DEVICES\n", .{});
        const count = c.soundio_output_device_count(self.soundio);
        return self.getNativeDevices(buffer, count, c.soundio_get_output_device);
    }

    pub fn startCapture(self: *Devices, config: *ConnectContext) !void {
        ////////////////////////////////////////////////////////////////
        // Initialize the render and capture devices
        const capture = config.capture_device;
        const sio_capture = @intToPtr(*c.struct_SoundIoDevice, capture.platform_device);
        defer c.soundio_device_unref(sio_capture);
        print("Soundio capture name: {s}\n", .{sio_capture.name});

        const render = config.render_device;
        const sio_render = @intToPtr(*c.struct_SoundIoDevice, render.platform_device);
        defer c.soundio_device_unref(sio_render);
        print("Soundio render name: {s}\n", .{sio_render.name});

        const latency = DefaultLatencyMs;
        const buffer_latency = DefaultBufferMs;

        ////////////////////////////////////////////////////////////////
        // Configure Channel Layout
        // Find the first supported layout between the capture and render devices.

        var layout = sio_capture.current_layout;

        // c.soundio_device_sort_channel_layouts(sio_capture);
        // c.soundio_device_sort_channel_layouts(sio_render);
        // var layout = c.soundio_best_matching_channel_layout(sio_render.layouts, sio_render.layout_count, sio_capture.layouts, sio_capture.layout_count);
        // if (layout == null) {
        //     return AudError.StreamLayoutFailure;
        // }
        // TODO(lbayes): Figure out how to trace an identifier for layout
        print("Soundio using layout channel_count: {}\n", .{layout.channel_count});

        ////////////////////////////////////////////////////////////////
        // Configure Sample Rate
        // Find the first supported sample rate between the capture and render devices.
        var sample_rate: c_int = undefined;
        for (PrioritizedSampleRates[0..]) |rate| {
            sample_rate = @intCast(c_int, rate);
            if (c.soundio_device_supports_sample_rate(sio_capture, sample_rate) and
                c.soundio_device_supports_sample_rate(sio_render, sample_rate))
            {
                break;
            }
        }
        print("Soundio using sample_rate: {}\n", .{sample_rate});

        ////////////////////////////////////////////////////////////////
        // Configure Format
        var format: c.SoundIoFormat = c.enum_SoundIoFormat.Invalid;
        for (PrioritizedFormats[0..]) |p_format| {
            format = @intToEnum(c.SoundIoFormat, p_format);
            if (format != c.enum_SoundIoFormat.Invalid and
                c.soundio_device_supports_format(sio_capture, format) and
                c.soundio_device_supports_format(sio_render, format))
            {
                break;
            }
        }

        if (format == c.enum_SoundIoFormat.Invalid) {
            return AudError.FormatFailure;
        }
        print("Soundio using format: {}\n", .{format});

        ////////////////////////////////////////////////////////////////
        // Configure In Stream
        var in_stream: *c.struct_SoundIoInStream = c.soundio_instream_create(sio_capture);
        // TODO(lbayes): Figure out if we've got a null in_stream?
        defer c.soundio_instream_destroy(in_stream);

        in_stream.software_latency = latency;
        in_stream.format = format;
        in_stream.sample_rate = sample_rate;
        in_stream.layout = layout;

        in_stream.read_callback = sioCaptureCallback;
        in_stream.error_callback = sioCaptureErrorCallback;
        in_stream.overflow_callback = sioCaptureOverflowCallback;

        in_stream.userdata = config;

        var err = c.soundio_instream_open(in_stream);
        if (err != 0) {
            return AudError.StreamOpenFailure;
        }
        print("Soundio opened instream\n", .{});

        ////////////////////////////////////////////////////////////////
        // Configure Out Stream
        var out_stream: *c.struct_SoundIoOutStream = c.soundio_outstream_create(sio_render);
        // TODO(lbayes): Figure out if we've got a null out_stream?
        // defer c.soundio_outstream_destroy(out_stream);

        out_stream.software_latency = latency;
        out_stream.format = format;
        out_stream.sample_rate = sample_rate;
        out_stream.layout = layout;

        out_stream.write_callback = sioRenderCallback;
        out_stream.error_callback = sioRenderErrorCallback;
        out_stream.underflow_callback = sioRenderUnderflowCallback;

        out_stream.userdata = config;

        err = c.soundio_outstream_open(out_stream);
        if (err != 0) {
            return AudError.StreamOpenFailure;
        }
        print("Soundio opened outstream\n", .{});

        ////////////////////////////////////////////////////////////////
        // Create RingBuffer
        // const capacity: c_int = in_stream.bytes_per_frame * buffer_latency * sample_rate * 2;
        // const c_capacity: c_int = in_stream.bytes_per_frame * buffer_latency * sample_rate * 2;
        // const capacity: usize = @intCast(usize, c_capacity);

        // TODO(lbayes): Fix below. Got an LLVM IR error when attempting to intCast this at runtime.
        const capacity: usize = 716800000;
        print("Soundio ring_buffer capacity bytes: {}\n", .{capacity});
        var buffer = try RingBuffer(u8).init(self.allocator, capacity);
        config.buffer = buffer;

        // var ring_buffer = c.soundio_ring_buffer_create(self.soundio, capacity);
        // Store the ring_buffer pointer on the context
        // config.platform_buffer = @ptrToInt(ring_buffer);

        // var render_buffer = c.soundio_ring_buffer_write_ptr(ring_buffer);
        // TODO(lbayes): Figure if we need to advance the write ptr at start
        // TODO(lbayes): Make the ring_buffer accessible to callbacks
        print("soundio ring_buffer created\n", .{});

        ////////////////////////////////////////////////////////////////
        // Cross the streams!
        c.soundio_flush_events(self.soundio);
        print(">>>>>>>>>>>>>> Soundio render name: {s}\n", .{sio_render.name});

        // TODO(lbayes): THE FOLLOWING LINE SEGFAULTS!
        // err = c.soundio_outstream_start(out_stream);
        // if (err != 0) {
        //     return AudError.StreamStartFailure;
        // }
        // print("Soundio started outstream\n", .{});

        err = c.soundio_instream_start(in_stream);
        if (err != 0) {
            return AudError.StreamStartFailure;
        }

        print("Waiting for stream to close\n", .{});

        while (config.is_active) {
            c.soundio_wait_events(self.soundio);
            time.sleep(40 * time.ns_per_ms);
        }
    }

    fn sioRenderCallback(c_outstream: [*c]c.struct_SoundIoOutStream, frame_count_min: c_int, frame_count_max: c_int) callconv(.C) void {
        print("SIO RENDER CB\n", .{});
        // const outstream: *c.struct_SoundIoOutStream = @as(*c.struct_SoundIoOutStream, c_outstream);
        // const context: *ConnectContext = @ptrCast(*ConnectContext, outstream.userdata);

        // print("soundio render: {} {} {}\n", .{ context.is_active, frame_count_min, frame_count_max });
    }

    fn sioRenderErrorCallback(instream: [*c]c.struct_SoundIoOutStream, err: c_int) callconv(.C) void {
        print("ERROR: soundio render: {}\n", .{err});
    }

    fn sioRenderUnderflowCallback(outstream: [*c]c.struct_SoundIoOutStream) callconv(.C) void {
        print("UNDERFLOW: soundio render\n", .{});
    }

    fn sioCaptureCallback(c_instream: [*c]c.struct_SoundIoInStream, frame_count_min: c_int, frame_count_max: c_int) callconv(.C) void {
        const instream: *c.struct_SoundIoInStream = @as(*c.struct_SoundIoInStream, c_instream);
        const context: *ConnectContext = @ptrCast(*ConnectContext, instream.userdata);
        // const ring_buffer = @intToPtr(*c.struct_SoundIoRingBuffer, context.platform_buffer);

        print("soundio capture: {} {} {}\n", .{ context.is_active, frame_count_min, frame_count_max });
        print("soundio bytes_per_frame: {}\n", .{instream.bytes_per_frame});

        // var frame: c_int = 0;
        // while (frame < frame_count_max) {
        //     print(
        //     frame += 1;
        // }
        // _ = std.c.printf("HELLO WORLD\n");
        // print("SIO READ CB\n", .{});
    }

    fn sioCaptureErrorCallback(instream: [*c]c.struct_SoundIoInStream, err: c_int) callconv(.C) void {
        print("ERROR: soundio capture: {}\n", .{err});
    }

    fn sioCaptureOverflowCallback(instream: [*c]c.struct_SoundIoInStream) callconv(.C) void {
        print("OVERFLOW: soundio capture\n", .{});
    }
};

test "Soundio Devices is instantiable" {
    var api = try Devices.init(talloc);
    defer api.deinit();

    try expectEqualStrings("soundio", api.info());
}

// NOTE(lbayes): This test will only pass on my workstation
// with particular devices connected. It is only useful for development purposes
// test "Soundio getDevices returns list" {
//     var api = try Devices.init(talloc);
//     defer api.deinit();
//
//     var buff: [common.MAX_DEVICE_COUNT]Device = undefined;
//     const devices = try api.getDevices(&buff, Direction.Render);
//
//     for (devices) |dev| {
//         print("DEVI: {s}\n", .{dev.name});
//     }
//     try expectEqual(devices.len, 5);
// }

File: cee/test/fakes/soundio/soundio.c
#include "soundio.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

//---------------------------------------------------------
// soundio
//---------------------------------------------------------

static struct SoundIo *context;
static bool use_fake_context = false;
static int fake_connect_status = 0;
static SoundIoFakeFunctionNames fake_last_function = FunctionNameNone;
static struct SoundIoDevice *fake_devices = {0};
static size_t fake_devices_count = 0;

void set_fake_devices(struct SoundIo *sio, enum SoundIoDeviceAim aim, unsigned int count, ...) {
  struct SoundIoDevice *devices = calloc(sizeof(struct SoundIoDevice), count);
  va_list vargs;
  va_start(vargs, count);
  for (unsigned int i = 0; i < count; ++i) {
    char *base_name = va_arg(vargs, char *);
    size_t base_count = strlen(base_name);
    
    char *name = calloc(base_count + 1, 1);
    strcpy(name, base_name);

    char *id = calloc(base_count + 4, 1);
    strcpy(id, base_name);
    strcat(id, "-id");

    devices[i].id = id;
    devices[i].name = name;
  }

  if (aim == SoundIoDeviceAimInput) {
    sio->fake_input_device_count = count;
    sio->fake_input_devices = devices;
  } else {
    sio->fake_output_device_count = count;
    sio->fake_output_devices = devices;
  }
  va_end(vargs);
}

SOUNDIO_EXPORT struct SoundIo *soundio_create(void) {
  if (!use_fake_context) {
    struct SoundIo *s = malloc(sizeof(struct SoundIo));
    context = s;
    context->fake_input_device_count = 0;
    context->fake_output_device_count = 0;
    return s;
  }
  use_fake_context = false;
  fake_last_function = FunctionNameCreate;
  return context;
}

void soundio_fake_create_returns(struct SoundIo *sio) {
  use_fake_context = true;
  context = sio;
  if (sio != NULL) {
    sio->fake_input_device_count = 0;
    sio->fake_output_device_count = 0;
  }
}

SOUNDIO_EXPORT int soundio_connect(struct SoundIo *soundio) {
  int status = fake_connect_status;
  fake_connect_status = 0;
  fake_last_function = FunctionNameConnect;
  return status; 
}

SOUNDIO_EXPORT SoundIoFakeFunctionNames soundio_fake_get_last_function_name(void) {
  return fake_last_function;
}

void soundio_fake_connect_returns(int status) {
  fake_connect_status = status;
}

SOUNDIO_EXPORT int soundio_connect_backend(struct SoundIo *soundio, enum SoundIoBackend backend) {
  int result = soundio_connect(soundio);
  fake_last_function = FunctionNameConnectBackend;
  return result;
}

SOUNDIO_EXPORT void soundio_flush_events(struct SoundIo *soundio) {

}

SOUNDIO_EXPORT void soundio_disconnect(struct SoundIo *soundio) {
}

SOUNDIO_EXPORT void soundio_sort_channel_layouts(struct SoundIoChannelLayout *layouts, int layout_count) {
}

SOUNDIO_EXPORT const char *soundio_strerror(int error) {
  return "fake-error";
}

void free_devices(struct SoundIoDevice *devices, int count) {
  if (count > 0) {
    for (int i = 0; i < count; i++) {
      free(devices[i].id);
      free(devices[i].name);
    }
    free(devices);
  }
}

SOUNDIO_EXPORT void soundio_destroy(struct SoundIo *sio) {
  free_devices(sio->fake_input_devices, sio->fake_input_device_count);
  free_devices(sio->fake_output_devices, sio->fake_output_device_count);
  free(sio);
}

//---------------------------------------------------------
// devices
//---------------------------------------------------------

SOUNDIO_EXPORT bool soundio_device_supports_format(struct SoundIoDevice *device,
    enum SoundIoFormat format) {
  return false;
}

SOUNDIO_EXPORT bool soundio_device_supports_layout(struct SoundIoDevice *device,
    const struct SoundIoChannelLayout *layout) {
  return false;
}

SOUNDIO_EXPORT bool soundio_device_supports_sample_rate(struct SoundIoDevice *device,
    int sample_rate) {
  return false;
}

SOUNDIO_EXPORT void soundio_device_sort_channel_layouts(struct SoundIoDevice *device) {
}

SOUNDIO_EXPORT const struct SoundIoChannelLayout *soundio_best_matching_channel_layout(
    const struct SoundIoChannelLayout *preferred_layouts, int preferred_layout_count,
    const struct SoundIoChannelLayout *available_layouts, int available_layout_count) {
  return NULL;
}

SOUNDIO_EXPORT struct SoundIoDevice *soundio_get_input_device(struct SoundIo *soundio, int index) {
  return &soundio->fake_input_devices[index];
}

SOUNDIO_EXPORT struct SoundIoDevice *soundio_get_output_device(struct SoundIo *soundio, int index) {
  return &soundio->fake_output_devices[index];
}

SOUNDIO_EXPORT int soundio_default_input_device_index(struct SoundIo *soundio) {
  return 0; 
}

SOUNDIO_EXPORT int soundio_default_output_device_index(struct SoundIo *soundio) {
  return 0;
}

SOUNDIO_EXPORT int soundio_input_device_count(struct SoundIo *soundio) {
  return soundio->fake_input_device_count;
}

SOUNDIO_EXPORT int soundio_output_device_count(struct SoundIo *soundio) {
  return soundio->fake_output_device_count;
}

SOUNDIO_EXPORT void soundio_device_unref(struct SoundIoDevice *device) {
}

//---------------------------------------------------------
// out_stream
//---------------------------------------------------------
SOUNDIO_EXPORT struct SoundIoOutStream *soundio_outstream_create(struct SoundIoDevice *device) {
  return NULL;
}

SOUNDIO_EXPORT int soundio_outstream_open(struct SoundIoOutStream *outstream) {
  return -1;
}

SOUNDIO_EXPORT int soundio_outstream_start(struct SoundIoOutStream *outstream) {
  return -1;
}

SOUNDIO_EXPORT int soundio_outstream_begin_write(struct SoundIoOutStream *outstream,
    struct SoundIoChannelArea **areas, int *frame_count) {
  return -1;
}

SOUNDIO_EXPORT int soundio_outstream_end_write(struct SoundIoOutStream *outstream) {
  return -1;
}

SOUNDIO_EXPORT void soundio_outstream_destroy(struct SoundIoOutStream *outstream) {
}

//---------------------------------------------------------
// in_stream
//---------------------------------------------------------
SOUNDIO_EXPORT struct SoundIoInStream *soundio_instream_create(struct SoundIoDevice *device) {
  return NULL;
}

SOUNDIO_EXPORT int soundio_instream_open(struct SoundIoInStream *instream) {
  return -1;
}

SOUNDIO_EXPORT int soundio_instream_start(struct SoundIoInStream *instream) {
  return -1;
}

SOUNDIO_EXPORT int soundio_instream_begin_read(struct SoundIoInStream *instream,
    struct SoundIoChannelArea **areas, int *frame_count) {
  return -1;
}

SOUNDIO_EXPORT int soundio_instream_end_read(struct SoundIoInStream *instream) {
  return -1;
}

SOUNDIO_EXPORT void soundio_instream_destroy(struct SoundIoInStream *instream) {
}



//---------------------------------------------------------
// ring_buffer
//---------------------------------------------------------
SOUNDIO_EXPORT struct SoundIoRingBuffer *soundio_ring_buffer_create(struct SoundIo *soundio, int requested_capacity) {
  return NULL;
}

SOUNDIO_EXPORT void soundio_ring_buffer_destroy(struct SoundIoRingBuffer *ring_buffer) {
}

SOUNDIO_EXPORT int soundio_ring_buffer_capacity(struct SoundIoRingBuffer *ring_buffer) {
  return 0;
}

SOUNDIO_EXPORT int soundio_ring_buffer_fill_count(struct SoundIoRingBuffer *ring_buffer) {
  return 0;
}

SOUNDIO_EXPORT int soundio_ring_buffer_free_count(struct SoundIoRingBuffer *ring_buffer) {
  return 0;
}

SOUNDIO_EXPORT char *soundio_ring_buffer_read_ptr(struct SoundIoRingBuffer *ring_buffer) {
  return "a";
}

SOUNDIO_EXPORT void soundio_ring_buffer_advance_read_ptr(struct SoundIoRingBuffer *ring_buffer, int count) {
}

SOUNDIO_EXPORT char *soundio_ring_buffer_write_ptr(struct SoundIoRingBuffer *ring_buffer) {
  return "a";
}

SOUNDIO_EXPORT void soundio_ring_buffer_advance_write_ptr(struct SoundIoRingBuffer *ring_buffer, int count) {
}


